php_buffer.h

```diff
@@ -1,42 +1,90 @@
-/*
-  +----------------------------------------------------------------------+
-  | PHP Version 5                                                        |
-  +----------------------------------------------------------------------+
-  | Copyright (c) 1997-2012 The PHP Group                                |
-  +----------------------------------------------------------------------+
-  | This source file is subject to version 3.01 of the PHP license,      |
-  | that is bundled with this package in the file LICENSE, and is        |
-  | available through the world-wide-web at the following url:           |
-  | http://www.php.net/license/3_01.txt                                  |
-  | If you did not receive a copy of the PHP license and are unable to   |
-  | obtain it through the world-wide-web, please send a note to          |
-  | license@php.net so we can mail you a copy immediately.               |
-  +----------------------------------------------------------------------+
-  | Author:                                                              |
-  +----------------------------------------------------------------------+
-*/
-
-/* $Id$ */
-
 #ifndef PHP_BUFFER_H
 #define PHP_BUFFER_H
 
+#define BUFFER_VERSION "0.2"
+
 extern zend_module_entry buffer_module_entry;
 #define phpext_buffer_ptr &buffer_module_entry
 
 #ifdef PHP_WIN32
-#	define PHP_BUFFER_API __declspec(dllexport)
+#      define PHP_BUFFER_API __declspec(dllexport)
 #elif defined(__GNUC__) && __GNUC__ >= 4
-#	define PHP_BUFFER_API __attribute__ ((visibility("default")))
+#      define PHP_BUFFER_API __attribute__ ((visibility("default")))
 #else
-#	define PHP_BUFFER_API
+#      define PHP_BUFFER_API
 #endif
 
 #ifdef ZTS
 #include "TSRM.h"
 #endif
+#pragma once
+
+#if defined(ZTS) && defined(COMPILE_DL_BUFFER)
+ZEND_TSRMLS_CACHE_EXTERN();
+#endif
+
+typedef struct _buffer_object {
+  void *buffer;
+  size_t length;
+
+  zend_object std;
+} buffer_object;
+
+typedef enum _buffer_view_type {
+	buffer_view_int8,
+	buffer_view_uint8,
+	buffer_view_int16,
+	buffer_view_uint16,
+	buffer_view_int32,
+	buffer_view_uint32,
+	buffer_view_float,
+	buffer_view_double
+} buffer_view_type;
+
+typedef struct _buffer_view_object {
+	zval buffer_zval;
+
+	union {
+		int8_t   *as_int8;
+		uint8_t  *as_uint8;
+		int16_t  *as_int16;
+		uint16_t *as_uint16;
+		int32_t  *as_int32;
+		uint32_t *as_uint32;
+		float    *as_float;
+		double   *as_double;
+	} buf;
+
+	size_t offset;
+	size_t length;
+
+	/* For Iterator methods */
+	size_t current_offset;
```

Comment:
After this, I want to do a migration to PHP 8, maybe leave that change for then? Otherwise I can give this a try now, just let me know.

Something else we need for Rubix is for Buffer to implement the Countable interface, that's also next for me but I hope the current migration is solid before adding changes.